state/endpoint_bindings.go

```diff
@@ -200,14 +241,84 @@ func updateEndpointBindingsOp(st *State, key string, givenMap map[string]string,
 
 	updateOp := txn.Op{
 		C:      endpointBindingsC,
-		Id:     key,
+		Id:     a.globalKey(),
 		Update: bson.M{"$set": bson.M{"bindings": escaped}},
 	}
 	if existingMap != nil {
 		// Only assert existing haven't changed when they actually exist.
 		updateOp.Assert = bson.D{{"txn-revno", txnRevno}}
 	}
-	return updateOp, nil
+
+	return append(ops, updateOp), nil
+}
+
+// validateEndpointBindingsForMachines checks whether the required space
+// assignments are actually feasible given the network configuration settings
+// of the machines where application units are already running.
+func validateEndpointBindingsForMachines(a *Application, oldBindings, newBindings map[string]string) error {
```

Comment:
Passing comment:

I can see that it is hard to do for this particular piece of logic, because `updateEndpointBindingsOps` is called from `Application.changeCharmOps`, but these kind of cross-cutting concerns are where we should be extracting a `ModelOperation` to keep the concerns isolated - charms/applications/machines are all woven together in state here.